plutus-core/src/Language/PlutusCore/Parser.y

```diff
@@ -190,22 +180,100 @@ getStaticBuiltinName = \case
     "ifThenElse"               -> Just IfThenElse
     _                          -> Nothing
 
-mkBuiltin :: a -> T.Text -> Term TyName Name uni a
-mkBuiltin loc ident = 
+
+--- Parsing built-in types and constants ---
+-- Most of this stuff will be moved into a Parsable class at some point
+
+tagOfTyName :: T.Text -> Maybe Int
+tagOfTyName = \case 
+  "bool"       -> Just $ tagOf DefaultUniBool
+  "bytestring" -> Just $ tagOf DefaultUniByteString
+  "char"       -> Just $ tagOf DefaultUniChar
+  "integer"    -> Just $ tagOf DefaultUniInteger
+  "string"     -> Just $ tagOf DefaultUniString
+  "unit"       -> Just $ tagOf DefaultUniUnit
+  _ -> Nothing
+
+mkBuiltinType
+  :: (DefaultUni <: uni, Closed uni)
+    => AlexPosn -> T.Text -> Parse (Type TyName uni AlexPosn)
+mkBuiltinType tyloc tyname = case tagOfTyName tyname of
+  Nothing -> throwError $ UnknownBuiltinType tyloc tyname
+  Just tag ->
+    case uniAt tag of
+          Nothing -> throwError $ UnknownBuiltinType tyloc tyname
+          Just ty -> pure $ TyBuiltin tyloc ty
+  
+
+-- FIXME: I can't see how to do this without dispatching on the type 
+-- because we're doing non-uniform things.
+mkBuiltinConstant
+  :: DefaultUni <: uni
+  => AlexPosn -> T.Text -> AlexPosn -> T.Text -> Parse (Term TyName Name uni AlexPosn)
+mkBuiltinConstant tyloc tyname litloc lit  = do
+    val <-
+          case tyname of
+            "bool"       -> readConstant @ Bool    litloc lit
+            "bytestring" -> parseByteString tyname litloc lit
+            "char"       -> readConstant @ Char    litloc lit 
+            "integer"    -> readConstant @ Integer litloc lit 
+            "string"     -> readConstant @ String  litloc lit
+            "unit"       -> readConstant @ ()      litloc lit 
+            _            -> throwError $ UnknownBuiltinType tyloc tyname
+    case val of
+         Nothing -> throwError $ InvalidBuiltinConstant litloc lit tyname
+         Just v -> pure v
+
+readConstant
+    :: forall t a uni. (Read t, uni `Includes` t)
+    => AlexPosn -> T.Text -> Parse (Maybe (Term TyName Name uni AlexPosn))
+readConstant loc lit = pure $ fmap (mkConstant loc) $ (readMaybe @ t (T.unpack lit))
```

Comment:
Can you just drop the `pure` and the `Parse` on the return type?

Reply:
Well-spotted. Must have been left over from an earlier version.